x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java

```diff
@@ -73,13 +117,75 @@ ActionListener<Void> acquireAvoid() {
+     * compute request to a remote cluster.
+     * @param clusterAlias clusterAlias of cluster receiving the remote compute request
+     * @return Listener that will fill in all metadata from to remote cluster into the
+     *         {@link EsqlExecutionInfo}  for the clusterAlias cluster.
+     */
+    ActionListener<ComputeResponse> acquireCCSCompute(String clusterAlias) {
+        assert clusterAlias != null : "Must provide non-null cluster alias to acquireCompute";
+        assert esqlExecutionInfo != null : "When providing cluster alias to acquireCompute, EsqlExecutionInfo must not be null";
+        return acquireAvoid().map(resp -> {
+            responseHeaders.collect();
+            var profiles = resp.getProfiles();
+            if (profiles != null && profiles.isEmpty() == false) {
+                collectedProfiles.addAll(profiles);
+            }
+            esqlExecutionInfo.swapCluster(
+                clusterAlias,
+                (k, v) -> new EsqlExecutionInfo.Cluster.Builder(v)
+                    // MP TODO: if we get here does that mean that the remote search is finished and was SUCCESSFUL?
+                    // for now ESQL doesn't return partial results, so set status to SUCCESSFUL
+                    .setStatus(EsqlExecutionInfo.Cluster.Status.SUCCESSFUL)
+                    .setTook(resp.getTook())
+                    .setTotalShards(resp.getTotalShards())
+                    .setSuccessfulShards(resp.getSuccessfulShards())
+                    .setSkippedShards(resp.getSkippedShards())
+                    .setFailedShards(resp.getFailedShards())
+                    .build()
+            );
+            return null;
+        });
+    }
+
+    /**
+     * Acts like {@code acquireCompute} for handling the response(s) form the runComputeOnDataNodes
+     * phase. Per-cluster took time is recorded in the {@link EsqlExecutionInfo}.
+     * @param clusterAlias remote cluster alias the data node compute is running on
+     */
+    ActionListener<ComputeResponse> acquireComputeForDataNodes(String clusterAlias, Configuration configuration) {
+        assert clusterAlias != null : "Must provide non-null cluster alias to acquireCompute";
         return acquireAvoid().map(resp -> {
             responseHeaders.collect();
             var profiles = resp.getProfiles();
             if (profiles != null && profiles.isEmpty() == false) {
                 collectedProfiles.addAll(profiles);
             }
+            long tookTimeMillis = System.currentTimeMillis() - configuration.getQueryStartTimeMillis();
+            TimeValue tookOnDataNode = new TimeValue(tookTimeMillis);
+            esqlExecutionInfo.swapCluster(clusterAlias, (k, v) -> {
+                if (v.getTook() == null || v.getTook().millis() < tookOnDataNode.millis()) {
+                    return new EsqlExecutionInfo.Cluster.Builder(v).setStatus(EsqlExecutionInfo.Cluster.Status.SUCCESSFUL)
```

Comment:
This doesn't seem right. It looks like we received a response from only one node, rather than from the entire cluster.

Reply:
Yes, good catch. I forgot to remove the setting of SUCCESSFUL status here when doing code cleanup. That status gets properly set elsewhere. I removed that in [this commit](https://github.com/elastic/elasticsearch/pull/112472/commits/09fa22441a860e4e22b1983382aa2138afd53406), but left in the took time calculation code.